### ISSUE TITLE
MIDI playback in hard RT context

### ISSUE THREAD

**Issue by jpcima** [Wohlstand/libADLMIDI#65]
In a softsynth plugin, MIDI playback can happen in the same thread as audio processing.
adlmidi_midiplay must be modified for hard-RT to avoid all kinds of malloc.

To give an example, `activenotes` is stored in a std::map of uint8_t.
In a std::map `malloc` is required for each insertion, ie. for every MIDI note on.
Midi notes being limited to 127, it would be a better idea to use flat array storage.

Please assign me the issue.

**Comment by jpcima** [comment]
There are multiple strategies to solve the problem of note keeping.
(In all cases I believe the array might be in fact faster because of low element count and cache)

1. the most obvious may be to replace `map` with `vector` reserved with 128 cells. The positive is that I only have to iterate up to size, but removal will take O(n).
2. Just make it an array of 128 and iterate everything.

I guess I safely ignore the consideration of ordering of the note array? originally std::map is ordered by ascending note number.

**Comment by Wohlstand** [comment]
Third strategy:
* Create 128 array which will index all notes.
* Create `vector` which will contain a chain of `playing indexes` to iterate them.

Ow! just now I have found that each `NoteInfo` contains another map `std::map<uint16_t /*adlchn*/, Phys> phys;` which refers using physical chip(s) channels (one two-op voice, two two-op voices or two pieces of four-op voice). So, I'll check out an ability to replace it with fixed-size array and the count value.

**Comment by jpcima** [comment]
Yes I'm aware of this other pesky little map.
I have already made its acquaintance in original Bisqwit code.

**Comment by jpcima** [comment]
I made the hard-realtime branch. This is a simple transformation of `activenotemap` into a fixed array.
It could be organized a bit better for sure.
Do you approve this?

https://github.com/jpcima/libADLMIDI/commit/21fd71af4394f88522630d9a900dc1d8d8ab8625

**Comment by Wohlstand** [comment]
I have continued stuff, and I removed the "phys" map with replacing it with two-element arrays, that also need to test out for any possible issues with it. Also, I gonna to move MIDI Sequencer which loads and plays MIDI files in-home of library into separated file, then into separated class that will be possible to use in other libraries separately.

**Comment by jpcima** [comment]
Nice! I will look at the code that you have done, and try it. Thanks.

I know there remains the `users` map, from `AdlChannel`.
I have not completely understood the full purpose of this map. Is it a history of the current and old notes on the channel?
Do you know if this one should this one be bounded to 2 as well? or more?

> Also, I gonna to move MIDI Sequencer which loads and plays MIDI files in-home of library into separated file

Cool I have a benchark of quality of MIDI file implementations. I could include ADLMIDI for a test.

**Comment by jpcima** [comment]
Well this is the annoying point.. I don't think there is a bound to the size of this map.
Just looking at code, `users` is apparently not limited in growth.
Allocating everything is out of the question. (AdlChannels * MidiChannels * Notes)

Instead, maybe it should be bounded by a generous max polyphony value? (say, for example 32)
A question is, what happens if the user overflowed the polyphony amount?
I can of think of multiple answers, but I don't yet know about the adverse effects which may happen.

**Comment by Wohlstand** [comment]
32 seems a small range, probably 128 or 256 slots. And yeah, removal from a middle will need to do the shift of part of array data, even adding is always going into the last cell. Anyway, as an a crazy idea, I have got an idea of statically allocated ring buffer. Which means, it will contain pointers NOT to far-placed heap cells, but will point neighbors, and "free" cells are must be remembered in separated array to easily begin usage of it. Pointers will help to quickly skip free cells to faster iterate the buffer.

**Comment by jpcima** [comment]
Like an intrusive list?

**Comment by Wohlstand** [comment]
I think, yeah, and it's purpose to allow storing into it, quick iteration, removal from any position, and everything must be inside of the statical buffer range! :fox_face:

**Comment by Wohlstand** [comment]
Also, when note got off, it still be in the list until time out to don't break releasing sound.

**Comment by jpcima** [comment]
It does not bother me. Only the GUI thread will parse WOPL data.
From there it will transmit the instruments to RT by appropriate APIs.

**Comment by Wohlstand** [comment]
Then OK, and this means, when you loading a bank file, temporarily pause rt while you are replacing back data, and then re-enable it again.

**Comment by jpcima** [comment]
>when you loading a bank file, temporarily pause rt while you are replacing back data, and then re-enable it again.

No. I will not stop the world during this process. I will send the banks from UI to RT over ring buffer. The ring buffer is lock free.
I will make the UI wait to send each bank in succession, while it'll wait for RT to consume the message queue in parallel.

I wish for RT not to require a lock to load its banks. Event if it's not the case in the present, there is perspective for a sysex protocol for RT to receive its banks by itself, which involves in no way the GUI.

**Comment by Wohlstand** [comment]
Then yeah, you even not needed for the LoadBank call as you will fill pre-allocated empty bank. Anyway, in case of multi-bank reallocation would be needed when reserved buffer will be overfilled while GM bank is always fixed size.

**Comment by jpcima** [comment]
>Anyway, in case of multi-bank reallocation would be needed when reserved

It is the reason why I ask you if I can preallocate my N bank slots in advance. I don't want this reallocation.
If there isn't room for more banks, I will be happy if you just return me a NULL.

**Comment by Wohlstand** [comment]
> It is the reason why I ask you if I can preallocate my N bank slots in advance. I don't want this reallocation.
If there isn't room for more banks, I will be happy if you just return me a NULL.

Okay then, that will be. So, when you will use `LoadBank` (but I'll stop to clear bank data and null it instead), it will allocate bank as needed, but for the alternative I'll give the ability to pre-allocate N banks as we already discussed. So, I'll add the item into the tasks list of #59 :wink:

**Comment by Wohlstand** [comment]
Looks like this thing is also completed.

**Comment by jpcima** [comment]
It is!

**Comment by Wohlstand** [comment]
And this also, closing! :wink:
